src/factories/workflow-specific/Restricted_PIM_Factory_v1.sol

```diff
@@ -0,0 +1,238 @@
+            bcProperties.initialIssuanceSupply,
+            realAdmin,
+            _msgSender()
+        );
+
+        // assign permissions to buy/sell from curve to admin
+        bytes32 curveAccess = FM_BC_Restricted_Bancor_Redeeming_VirtualSupply_v1(
+            fundingManager
+        ).CURVE_INTERACTION_ROLE();
+        FM_BC_Restricted_Bancor_Redeeming_VirtualSupply_v1(fundingManager)
+            .grantModuleRole(curveAccess, _msgSender());
+        // revoke minting rights from factory
+        issuanceToken.setMinter(address(this), false);
+
+        // revoke privileges from factory
+        _transferTokenOwnership(issuanceToken, realAdmin);
+        _transferWorkflowAdminRights(orchestrator, realAdmin);
+
+        emit IRestricted_PIM_Factory_v1.PIMWorkflowCreated(
+            address(orchestrator), address(issuanceToken), _msgSender()
+        );
+    }
+
+    /// @inheritdoc IRestricted_PIM_Factory_v1
+    function addFunding(address actor, address token, uint amount) external {
+        // records funding amount
+        fundings[_msgSender()][actor][token] += amount;
+        // sends amount from msg.sender to factory
+        IERC20(token).safeTransferFrom(_msgSender(), address(this), amount);
+
+        emit IRestricted_PIM_Factory_v1.FundingAdded(
+            _msgSender(), actor, token, amount
+        );
+    }
+
+    /// @inheritdoc IRestricted_PIM_Factory_v1
+    function withdrawFunding(address actor, address token, uint amount)
+        external
+    {
+        // checks if the requested amount is available
+        uint availableFunding = fundings[_msgSender()][actor][token];
+        if (amount > availableFunding) {
+            revert IRestricted_PIM_Factory_v1.InsufficientFunding(
+                availableFunding
+            );
+        }
+        // if so adjusts internal balancing
+        fundings[_msgSender()][actor][token] -= amount;
+        // and sends amount to msg sender
+        IERC20(token).safeTransfer(_msgSender(), amount);
+
+        emit IRestricted_PIM_Factory_v1.FundingRemoved(
+            _msgSender(), actor, token, amount
+        );
+    }
+
+    //--------------------------------------------------------------------------
+    // Internal Functions
+
+    function _manageInitialSupplies(
+        IBondingCurveBase_v1 fundingManager,
+        IERC20 collateralToken,
+        ERC20Issuance_v1 issuanceToken,
+        uint initialCollateralSupply,
+        uint initialIssuanceSupply,
+        address admin,
+        address actor
+    ) private {
+        uint availableFunding = fundings[admin][actor][address(collateralToken)];
+        if (availableFunding < initialCollateralSupply) {
+            revert IRestricted_PIM_Factory_v1.InsufficientFunding(
+                availableFunding
+            );
+        }
+
+        fundings[admin][actor][address(collateralToken)] -=
+            initialCollateralSupply;
+
+        // collateral token is paid for by the msg.sender
```

Comment:
This comment is confusing me slightly, as we are using the tokens that are deposited by the sponsor, not necessarily the msg.sender?

Reply:
yes, it should say either sponsor or factory

Reply:
yea, outdated comment is fixed